Sources/Ollama/Client.swift

```diff
@@ -326,8 +333,102 @@ extension Client {
         if let template {
             params["template"] = .string(template)
         }
-
+        
+        if toolDefinitions.count > 0 {
+            params["tools"] = try Value(toolDefinitions)
+        }
+        
         return try await fetch(.post, "/api/chat", params: params)
+        
+    }
+    
+    /// Generates the next message in a chat with a provided model.
+    ///
+    /// - Parameters:
+    ///   - model: The name of the model to use for the chat.
+    ///   - messages: The messages of the chat, used to keep a chat memory.
+    ///   - options: Additional model parameters as specified in the Modelfile documentation.
+    ///   - tools: Override the list of tools to include in this chat.
+    ///   - template: The prompt template to use (overrides what is defined in the Modelfile).
+    /// - Returns: A `ChatResponse` containing the generated message and additional information.
+    /// - Throws: An error if the request fails or the response cannot be decoded.
+    public func autoRunChat(
+        model: Model.ID,
+        messages: [Chat.Message],
+        options: [String: Value]? = nil,
+        tools: [Chat.ToolDefinition]? = nil,
+        template: String? = nil
+    )
+    async throws -> ([Chat.Message], ChatResponse)
+    {
+        let response: ChatResponse = try await chat(
+            model: model,
+            messages: messages,
+            options: options,
+            tools: tools,
+            template: template
+        )
+        
+        if let toolCalls = response.message.tool_calls
+        {
+            let (_, replyMessage) = try self.processToolCalls(toolCalls)
+            
+            var newMessages: [Chat.Message] = messages
+            newMessages.append(response.message)
+            newMessages.append(replyMessage)
+            
+            let response: ChatResponse = try await chat(
+                model: model,
+                messages: newMessages,
+                options: options,
+                tools: tools,
+                template: template
+            )
+            
+            return (newMessages, response)
+        }
+        else {
+            return (messages, response)
+        }
+    }
```

Comment:
I understand the sentiment, but at the same time when I think “tool support” this is the main component I think of. Without this it’s basically just a parsing the json into Swift structs.

when I think of tool support I think of providing functions to the LLM it can use in building its reply. So that I can ask a question, and get back a reply combining the LLM and the function.

Reply:
That's a decision only @mattt can make, but I think it's something that should be made clear.

Is this meant to be a fully fledged client in the Ollama "Get up and running with large language models" spirit? Or is it a thin _domain-transfer_ kind of utility wrapper?

Reply:
Given @mattt's comments I'm making a few changes to streamline this but keep most of the functionality (and adding @nicorichard's PR's to add async support).